### PR TITLE
Fullscreen enhancements (2563) & dynamic buttons

### LINKED ISSUE #2563: [CODE] Improve fullscreen system
**Description**
On a basic level, stop having fullscreen ClanGen be crunchy and gross. I am working on this currently, I thought I should drop in an issue to say "hey, this is being worked on".

Ideally, remove all the janky code throughout that handles fullscreen-vs-not and replace it with a less volatile system.

### Problems to Fix
- [x] `blit`ting backgrounds sends them to the shadow realm in the upper-left instead of the screenybit where they should be.
- [x] All that fullscreen code dotted around the codebase needs cleaning up/deleting
- [x] Determine & handle font size
- [x] Is it possible to toggle fullscreen without force-closing & manually restarting the game? surely it must be...?

### LINKED ISSUE #2603: [BUG] Continue Button Looks Weird
**Type**: (only select one)
- UI - anything to do with the user interface. maybe a button is not working, or the way some assets are formatted seems to be off. 

**Describe the bug**
The continue button looks off, there is space between "contin" and "ue" and the "ue" is bold. The space between the characters isn't uniform. I think all the main menu buttons should be procedurally generated when that's merged.

Probably going to be fixed in #2577.

**Grade**: (only select one)
- Unimportant - This bug has little to no impact on play. 

**Reproduce**
1. Start the game.
2. Look at the continue button.

**Commit # or Game Version Number**: 
0.11.2

**Screenshots**
problem2
problem

### LINKED ISSUE #2627: [BUG] Patrol ends unexpectedly when switching tab.
**Type**: (only select one)
- Code - general bugs related to how the game runs and responds to you. 

**Describe the bug**
Patrol ends unexpectedly when switching to a tab other than patrols.

**Grade**: (only select one)
- Unimportant - This bug has little to no impact on play. 

**Reproduce**

https://github.com/user-attachments/assets/25b6abcd-b3ab-4f15-9343-e6bb4951ff08

**Patrol ID + Outcome Print (if applicable)**:
image

**Commit # or Game Version Number**: 
7b48f10

### PR DESCRIPTION
**Pull request by j-gynn**
## About The Pull Request

So this started out as fullscreen enhancements and... spiralled.

**Summary of Key Changes**
- NEW: subclassed `UIManager` to allow it to control only a portion of the screen (therefore allowing the game to be offset to the center of the user's screen without having to rewrite every single line of code in the entire codebase).
- NEW: `UISurfaceImageButton`. This is a button subclass that accepts a dictionary of surfaces as the button images. It is totally theme-safe, can be reloaded & modified by the manager (update text etc.) and the surfaces stay intact.
- NEW: Procedurally generated buttons! There are a variety of flavours to suit the various places that buttons live in. It currently uses a kind of "three tile system", similar to a nine-tile system without a Y axis. This means buttons can be any length you want, but their height is locked - no multiline buttons currently. Here's a list of available buttons and what they look like:
	- SQUOVAL - the default button shape, you might prefer to call this "pill" or squircle... but squoval sparks joy so here we are
	image
	- ROUNDED_RECT - the squarish button
	image
	- DROPDOWN - Box edition
	image
	- MAINMENU - Th- the main menu buttons.
	image
	- MENU_LEFT, MENU_MIDDLE, MENU_RIGHT - the little units that make up a lot of our headers
	image
	- HORIZONTAL_TAB - used for tabbing horizontally, not sure what you want from me ;) Not used in many of the places it can go, mostly because I'm not super happy with the implementation. Text doesn't move during the state changes, so they look a little... flat. It's more noticeable with the vertical tab, which is why that has been entirely shelved for now.
        
        Note that buttons with unique styling or unique elements (blood, scratches, etc.) are not covered in this rework as they will require a much more nuanced approach. I have also not included the "ladder dropdowns" as there are some that are multiple lines long and this approach can only handle buttons with one row of text.

- NEW: The ClanGen font has had a bit of a spring-clean. Adjustment of some of the letters and their spacing, plus the addition of several constructors for arrows of arbitrary length using Unicode.
	- `add_arrow` allows an arrow to be inserted into any text box using the ClanGen font. It can create an arrow in lengths from 2 to infinity in steps of 0.5, with an additional option at length 1. All this to say, you'll probably have to eyeball exactly how big to make them, but the `back_button` uses an arrow length of 3.
- Screen scaling is now relative to 1x screen dimensions rather than 2x. This is probably what accounts for 50% of the line changes in this PR, as you will note that every single dimension is now half of what it was.
- NEW: added/updated scaling functions to work with the aforementioned scaling changes.
	- `ui_scale` (formerly `scale`) - 90% of use cases want this. It takes a pygame.Rect and scales it accordingly - appropriately scaling negative values for offsets (as anchors can use those) but not scaling negative dimensions as that indicates dynamic scaling.
	- `ui_scale_offset` - This is essentially the first two values of a pygame.Rect. Scales negative values.
	- `ui_scale_dimensions` - This is essentially the last two values of a pygame.Rect. Does not scale negative values.
	- `ui_scale_value` - Acts like a single value of ui_scale_dimensions (DOES scale negatives). Try not to use this - I've found it has to be used most when there's some really cursed screen building going on.
	- `ui_scale_blit` - Special function to allow for blitting images to the background at the right place whilst accounting for any fullscreen offset.
- NEW: The background of a screen is now handled by a handful of functions owned by the superclass.
	- `add_bgs` accepts a dictionary of names and pygame.Surfaces and, optionally, a dictionary that may contain one or more matching names to the first dictionary - this contains the background for fullscreen mode. If one is not provided, the game will take the main background, stretch and blur it to fill the space (see images).
	- `set_bg` accepts one of the names from the dictionary and sets it as the active background. A value of None can be passed to it - if so, it automatically defaults to the appropriate default background for the current theme (light mode or dark mode).
	- `show_bg` is called every frame and is used to redraw the background. If in fullscreen, it will blit the fullscreen background too, but if it isn't visible then for performance reasons it isn't blitted. No need.
- Combined all the theming files into one file. This might annoy some of you, but there is no reason for them to be split up the way they were - it made it hard to find things you were looking for.
	- That said, anything themeish that needs scaling with the UI is stored in `resources/theme/fonts/master_screen_scale.json`. This is because the game will automagically scale this to the right size on demand and reload it when the user changes screen dimensions. It can't be part of the same file as otherwise things like border thicknesses and corner rounds would also get messed with.
	- To use this new theming setup, ensure that *both themes are named the same*. The UI scaling theme file will override the base in the event they contain the same keys (e.g. both contain a "dark_bg" colour - the UI scaling file's BG will be the one that displays).
- NEW: You can toggle between fullscreen and windowed mode without the game crashing to desktop now!
- NEW: ConfirmDisplayChanges window now appears when you swap display mode. It's a timeout function: if you don't close it within 5 seconds, the changes are reverted.
- NEW: Fixed windows.py to use the existing functionality of UIWindow to handle blocking input to other buttons. In other words, `game.switches["window_open"]` is no more. Going forward, ensure that `process_event` on any UIWindow concludes by `return super().process_event()`, thereby using the `is_blocking` functionality that they're built with.
- Moved the screen settings functions (newly-made `toggle_fullscreen` and `load_manager`) out of `game_essentials` and into `scripts/game_structure/screen_settings`. This is to facilitate the possibility of loading the UI sooner in the load order so that there's not that big black screen for like 5 seconds on starting the game. Not currently implemented but I really, really wanna do it.
- Made some fixes to some of the most confusing Screens layouts so that they'll work with dynamic scaling. I tried to leave things alone as much as possible, but... this was so, so much easier. Please use anchors, dynamic scaling and mathematics will thank you.
- Added try:except wrapping around the save_settings() function to remove the onus from the end-user. It now accepts the calling screen as an argument, which it can use to return the user to the start screen to display an error message.
- Updated `UIImageButton` to always have `@image_button` as the class ID if a string/None value is passed in. This can be overridden by passing in an ObjectID-type item if desired.
- Cleaned up some parts of the SettingsScreen with the tooltips over credits - these really need refactoring, it's kind of a mess.
- Added the `anchors` field to `UIModifiedScrollingContainer` and `IDImageButton` to take advantage of anchoring for dynamic screen positioning. This already exists on the superclass, I'm just exposing it again.
- Due to the way the ingame `toggle_fullscreen` works, I've had to move the menu_buttons generation to a function that can be called again. We basically kill everything to start again.

**Known Issues/Limitations**
- ClanGen font has become very thin. not sure why.
- I haven't been able to convert all eligible buttons to using the dynamic enbuttoning. 
  - Pronouns use the CatButton to store a bit of extra data in there.
  - Leader's Den does some quirky stuff with the object ID - fixed it for outsiders but not for other_clan
  - One of the ladders has to stay because of strange things with exile buttons that we cycle through.

## Why This Is Good For ClanGen

Fullscreen has languished a little bit and been undesirable to play on due to the "crunchiness". Hopefully this resolves that!

## Linked Issues

Closes #2563
Closes #2603
Closes #2627 

## Proof of Testing

image
(running on 125% zoom @ 1920 x 1080 - game running at 1x scale)
image
(running on 100% zoom @ 1920 x 1080 - game running at 1.5x scale)

[Link to the Discord testing channel](https://discord.com/channels/1003759225522110524/1260014395560300614)

## Changelog/Credits

Fullscreen rework & Dynamic Buttons Part 1

### CODE REVIEW

**Comment by scribblecrumb** [comment]
Event screen merge left you with some conflicts ;-; godspeed soldier

**Comment by j-gynn** [comment]
So, so many bugs. Please hold.

**Comment by archanyhm** [comment]
Is it possible to split the dynamic buttons and the fullscreen enhancements up or would that be a lot of work for you? I'd usually argue that those two things are different enough to be in separate PRs, but I don't want to impose disproportionate amounts of work here.

**Comment by j-gynn** [comment]
> Is it possible to split the dynamic buttons and the fullscreen enhancements up or would that be a lot of work for you? I'd usually argue that those two things are different enough to be in separate PRs, but I don't want to impose disproportionate amounts of work here.

Teeechnically yes, but only if you're willing to accept that the fullscreen changes will include quite a few visual artifacts. The fullscreen PR is blocked by/dependent on the dynamic buttons & boxes in order to not do Weird Things. It's something to do with how scaling works and if I'm honest I don't fully remember the specifics, but those strange 1-2 pixel gaps that sometimes appear between some scaled elements get fixed by the dynamic buttons.

**Comment by j-gynn** [comment]
Bug found: when you load an existing clan for the first time, EventsScreen doesn't correctly load the past moon's events. It seems to load the relationship events fine, and the next moon loads as expected. Investigating.

**Comment by j-gynn** [comment]
Buggo squished

File: resources/placements.json
Comment (scribblecrumb) [inline resources/placements.json:1]: Am I right that this was just an auto-reformat thing?  I'm not a huge fan of how it reformatted, it's kinda *harder* to read now.

Comment (j-gynn) [reply resources/placements.json:1]: Regrettably yes, I didn't make it do that. As far as I can tell this was some kind of autoformatter and yes - it is definitely less readable. If you like, I can go through and manually fix it and see if that fixes it on commit.

Comment (j-gynn) [reply resources/placements.json:1]: I think? I have fixed it @scribblecrumb

**Review by scribblecrumb — CHANGES_REQUESTED** [review]
You're so cool for figuring all this out and I'm super excited to see it PR'd!  That said, here's what I noticed while playtesting.

image

Looks like unavailable main menu buttons are squovals on one side and rounded_rect on the other.

image

Not sure what's happening with the scroll bar here (windowed mode), but the arrow buttons seem very weird.  Eventually they'll get replaced by our more customized scrollbar, but until I get to this page in the refactor it'd be good to try and get the "default" scrollbar looking closer to how it used to.

image

Fullscreen version has no arrows at all (or maybe the top of one at the bottom button?)

image

The 'f' in info is being cut off in windowed mode. (this seems to happen with all F's)

image

Slightly less cut off in fullscreen.

image

(fullscreen) The tabs look to be a single pixel too far to the left.

image

Not a huge fan of how the menu buttons changed visually (the bars between each button always being visible), I'd prefer if they could stay consistent with their old visual.  I know this is probably to avoid having *another* type of button, but <3 know that I appreciate your hard work here and that also someone (maybe me?) tried to change it to that a long time ago and it wasn't a popular change At All.  I can already hear the complaints echoing.  Additionally, the middle buttons have the shadow on all sides, while the end buttons only have shadow on top and right.  This should also be made consistent.  Patrol text also looks slightly offcenter to the right.

image

Tabs have vertically stretched text.

image

sprite inspector has a MASSIVE fav star, probably not how it's meant to be?

image

(fullscreen) sprite inspector page has no checkboxes in this mode.

image

(fullscreen) Conditions box is one pixel too high.

image
Some buttons are a darker grey color scheme?  We want to use the lighter one.

image

Self-explanatory

image

Hard to demonstrate with a screenshot, but the next/prev buttons seem to be layered below the image, which causes them to only trigger a hover state when hovering their top half.

image

No seasons tabs in fullscreen or windowed?

image

Choose random button should be squoval

image

Clan name should be left aligned with the clan info.  Additionally, it scooted? to the right a couple times when I was clicking through symbols?  Doesn't happen consistently though.

image

Something funky happening with the filter window.

Other things:
- When opening the warriors den, the focus buttons visibly blink from a shorter version to a longer version, should probably have another look at how these are created in screen switch
- Could we get a fullscreen toggle on the clan settings page as well?
- Would love it if the camp BG was *always* the fullscreen BG when playing.  Blank color BG is boring and not particularly immersive.  I also find it a bit jarring when moving between screens.

**Comment by scribblecrumb** [comment]
image
tooltip box here should be wider

image
Unknown Residence displays the wrong BG

image
Event list scrollbar is a bit fucked in both full and windowed
Event list buttons (cat buttons and profile buttons) don't switch to hover state properly (sidenote: it's so cool to see the profile buttons actually use the clangen font)

image
The lighter rectangles on the event list look to be just slightly the wrong color, they should be the same color as the light mode BG

image
Fullscreened patrol artwork is still crunchy.  Is this a limitation or can it be remedied?

image
Header clan name isn't vertically centered in fullscreen

image
uh oh (windowed)

image
UH OH  (windowed)

image
Reflexive entry line seems to be missing

I'm also noticing that our Prev/Next cat buttons aren't visually consistent across different screens (profile vs. rel page, for example).

**Comment by scribblecrumb** [comment]
There's a pretty significant lag when leaving the settings + info page, seems much more pronounced when in fullscreen

image
There seems to be a white outline around the camp cats while in fullscreen, is this just a limitation or can we prevent it somehow?

image
Still broken in windowed mode after using the F4 toggle.  Kinda does the opposite when using the toggle to go from windowed to full (box becomes too small)  If this is just something we have to deal with with that toggle, then so be it, but it's not too ideal for debugging purposes.  I'll say, this seems to be the only screen where that box breaks.  Choose mate uses the same box but doesn't have this issue, however Change mentor *does* have the same issue.  Relationships screen problem seems to be fixed.

Allegiances page seems to open with a scrollbar even when the list isn't long enough to need one.  Scrolling all the way up then makes the bar disappear.

image
We're overlapping a tad

**Review by scribblecrumb — APPROVED** [review]

**Pull request merged**